argo_jupyter_scheduler/executor.py

```diff
@@ -573,3 +663,149 @@ def create_job_record(
 
         session.add(job)
         session.commit()
+
+
+@script()
+def rename_files(db_url, job_definition_id, input_path, log_path, start_time):
+    import os
+
+    from jupyter_scheduler.orm import Job, create_session
+
+    from argo_jupyter_scheduler.utils import (
+        add_file_logger,
+        gen_default_html_path,
+        gen_default_output_path,
+        gen_html_path,
+        gen_output_path,
+        gen_timestamp,
+        setup_logger,
+    )
+
+    try:
+        # Sets up logging
+        logger = setup_logger("rename_files")
+        add_file_logger(logger, log_path)
+
+        # Gets start_time if not provided to generate file paths
+        if start_time is None:
+            db_session = create_session(db_url)
+            with db_session() as session:
+                q = (
+                    session.query(Job)
+                    .filter(Job.job_definition_id == job_definition_id)
+                    .order_by(Job.start_time.desc())
+                    .first()
+                )
+
+                # The current job id doesn't match the id in the staging area.
+                # Creates a symlink to make files downloadable via the web UI.
+                basedir = os.path.dirname(os.path.dirname(input_path))
+                old_dir = os.path.join(basedir, job_definition_id)
+                new_dir = os.path.join(basedir, q.job_id)
+                os.symlink(old_dir, new_dir)
+
+                start_time = q.start_time
+
+        start_time = gen_timestamp(start_time)
+
+        old_output_path = gen_default_output_path(input_path)
+        old_html_path = gen_default_html_path(input_path)
+
+        new_output_path = gen_output_path(input_path, start_time)
+        new_html_path = gen_html_path(input_path, start_time)
+
+        # Renames files
+        os.rename(old_output_path, new_output_path)
+        os.rename(old_html_path, new_html_path)
```

Comment:
Note: it's important that the output path is processed first. If an exception is raised in `main` while executing a notebook, an HTML file won't be created. Doing it this way allows for downloading the notebook that was executed by papermill via the web UI. It will have the exception trace as well. The fact that the HTML rename will fail here is not important. The whole workflow will be marked as Failed and output will be logged to a file.

This is an example of an _output_ file from a job that failed during papermill execution (step `main`), downloaded via the web UI:
<img width="987" alt="Screen Shot 2023-12-08 at 22 17 45" src="https://github.com/nebari-dev/argo-jupyter-scheduler/assets/10469885/ad909ba5-cc35-497d-b48a-856c0996a4b2">